### ISSUE TITLE
Cannot redirect to charge confirmation from embedded app

### ISSUE THREAD

**Issue by nsweeting** [Shopify/shopify_app#398]
Hi there,

First time using embedded app stuff. So bare with me. When attempting to create a charge and do the redirect to the confirmation url, I end up with a blank page within the embedded app.

My controller code looks like this:

```ruby
  def create
    charge = current_shop.recurring_application_charges.new
    if charge.save
      fullpage_redirect_to(charge.confirmation_url)
    else
      #handle errors
    end
  end
```

I then end up with a blank page in the app. The iframe source looks like this (link details modified):

```html
        <!DOCTYPE html>
        <html lang="en">
          <head>
            <meta charset="utf-8" />
            <base target="_top">
            <title>Redirecting…</title>
            <script type="text/javascript">

              // If the current window is the 'parent', change the URL by setting location.href
              if (window.top == window.self) {
                window.top.location.href = "https://shop.myshopify.com/admin/charges/1/confirm_recurring_application_charge?signature=code_is_here";

              // If the current window is the 'child', change the parent's URL with postMessage
              } else {
                normalizedLink = document.createElement('a');
                normalizedLink.href = "https://shop.myshopify.com/admin/charges/1/confirm_recurring_application_charge?signature=code_is_here";

                data = JSON.stringify({
                  message: 'Shopify.API.remoteRedirect',
                  data: { location: normalizedLink.href }
                });
                window.parent.postMessage(data, "https://");
              }

            </script>
          </head>
          <body>
          </body>
        </html>
```

I also end up with the following error in the console:

```
Uncaught DOMException: Failed to execute 'postMessage' on 'Window': Invalid target origin 'https://' in a call to 'postMessage'.
    at https://my-app.herokuapp.com/subscriptions:23:31
```

Any help would be appreciated.

Thanks

**Comment by PresleyWong** [comment]
I experience same issue here.  I tried use older fullpage_redirect_to definition as below and it works but thats not a solution...

`	def fullpage_redirect_to(url)
	    url_json = url.to_json
	    url_json_no_quotes = url_json.gsub(/\A"|"\Z/, '')

	    if ShopifyApp.configuration.embedded_app?
	      render inline: %Q(
	        <!DOCTYPE html>
	        <html lang="en">
	          <head>
	            <meta charset="utf-8" />
	            <base target="_top">
	            <title>Redirecting…</title>
	            <script type="text/javascript">
	              window.top.location.href = #{url_json};
	            </script>
	          </head>
	          <body>
	          </body>
	        </html>
	      )
	    else
	      redirect_to_with_fallback url
	    end
	end`

**Comment by edavis10** [comment]
I ran into this on Monday.

`fullpage_redirect_to` is using the `sanitized_shop_name` method which uses `sanitize_shop_param` which checks `params[:shop]` for the shop. If you call it from a page there that params isn't set, then `full_page_redirect_to` won't be able to create the url for `postMessage`. You can see that in this line of generated JS code, the https:// should be the full shop url:

```
                window.parent.postMessage(data, "https://");
```

In my apps `current_shop` refers to the currently logged in shop record so I patched `sanitized_shop_name` to use that in addition to the params.

``` ruby
class ApplicationController < ActionController::Base
  def sanitized_shop_name
    if current_shop.present?
      @sanitized_shop_name ||= ShopifyApp::Utils.sanitize_shop_domain(current_shop.shopify_domain)
    else
      @sanitized_shop_name ||= sanitize_shop_param(params)
    end
    @sanitized_shop_name
  end
end
```

**Comment by kevinhughes27** [comment]
cc @Hammadk

**Comment by Hammadk** [comment]
On option here is to replace [`window.parent.postMessage(data, "https://#{sanitized_shop_name}");`](https://github.com/Shopify/shopify_app/blob/master/lib/shopify_app/login_protection.rb#L89) with `window.parent.postMessage(data, "*");`

Replacing the target origin with `*` means that we won't need to worry about the shop at all. We haven't done this so far because this approach is [not recommended](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage). 

I think it should be ok because we are only redirecting to Shopify specific URLs where the merchant has to be logged in. Thoughts?

**Comment by kevinhughes27** [comment]
I think I prefer @edavis10's solution.

**Comment by Hammadk** [comment]
@nsweeting @PresleyWong This issue should be fixed in version 7.2.6. Thanks for the suggestion @edavis10
